# 문제 1

In [ ]:
# pip install selenium

In [84]:
# !pip install pandas

In [2]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from bs4 import BeautifulSoup
import pandas as pd
import time

In [5]:
url = 'https://www.starbucks.co.kr/store/store_map.do?disp=locale'
driver = webdriver.Chrome()
driver.get(url)

In [6]:
# 서울 지역 선택
seoul_select = driver.find_element(By.CSS_SELECTOR, '#container > div > form > fieldset > div > section > article.find_store_cont > article > article:nth-child(4) > div.loca_step1 > div.loca_step1_cont > ul > li:nth-child(1) > a')
seoul_select.click()

In [7]:
# 구 선택
all_select = driver.find_element(By.CSS_SELECTOR, '#mCSB_2_container > ul > li:nth-child(1) > a')
all_select.click()

In [8]:
html = driver.page_source
soup = BeautifulSoup(html, 'html.parser')

stores = soup.select('.quickSearchResultBoxSidoGugun > li')

for store in stores:
    print(store.get_text())

 역삼아레나빌딩   서울특별시 강남구 언주로 425 (역삼동)1522-3232 리저브 매장 2번
 논현역사거리   서울특별시 강남구 강남대로 538 (논현동)1522-3232 리저브 매장 2번
 신사역성일빌딩   서울특별시 강남구 강남대로 584 (논현동)1522-3232 리저브 매장 2번
 국기원사거리   서울특별시 강남구 테헤란로 125 (역삼동)1522-3232 리저브 매장 2번
 대치재경빌딩   서울특별시 강남구 남부순환로 2947 (대치동)1522-3232 리저브 매장 2번
 봉은사역   서울특별시 강남구 봉은사로 619 (삼성동)1522-3232 리저브 매장 2번
 압구정윤성빌딩   서울특별시 강남구 논현로 834 (신사동)1522-3232 리저브 매장 2번
 코엑스별마당   서울특별시 강남구 영동대로 513 (삼성동)1522-3232 리저브 매장 2번
 삼성역섬유센터R   서울특별시 강남구 테헤란로 518 (대치동)1522-3232 리저브 매장 2번
 압구정R   서울특별시 강남구 언주로 861 (신사동)1522-3232 리저브 매장 2번
 수서역R   서울특별시 강남구 광평로 281 (수서동)1522-3232 리저브 매장 2번
 양재강남빌딩R   서울특별시 강남구 남부순환로 2621 (도곡동)1522-3232 리저브 매장 2번
 선릉동신빌딩R   서울특별시 강남구 테헤란로 409 (삼성동)1522-3232 리저브 매장 2번
 봉은사로선정릉   서울특별시 강남구 봉은사로 446 (삼성동)1522-3232 리저브 매장 2번
 강남오거리   서울특별시 강남구 봉은사로2길 39 (역삼동)1522-3232 리저브 매장 2번
 스타필드코엑스몰R   서울특별시 강남구 영동대로 513 (삼성동) 코엑스 A106호1522-3232 리저브 매장 2번
 강남구청정문   서울특별시 강남구 학동로 419 (청담동)1522-3232 리저브 매장 2번
 도곡공원   서울특별시 강남구 도곡로 205 (역삼동)1522-3232 리저브 매장 2번
 강남R   서울특별시 강남구 

- BeautifulSoup으로 가져온 매장 리스트에서 매장명과 주소를 각각 추출.
- 매장명은 strong 태그에서, 주소는 p class="result_details" 태그에서 추출.
- 추출한 데이터는 리스트(store_names와 addresses)에 저장.

In [9]:
import pandas as pd

store_names = []  
addresses = []    

# 매장 정보 추출
for store in stores:
    name_tag = store.find('strong') 
    if name_tag:
        store_names.append(name_tag.get_text(strip=True))  # 공백 제거 후 저장
    else:
        store_names.append("매장명 없음")
    
    address_tag = store.find('p', class_='result_details')  
    if address_tag:
        addresses.append(address_tag.get_text(strip=True))  
    else:
        addresses.append("주소 없음")


In [10]:
seoul_starbucks_df = pd.DataFrame({
    '매장명': store_names,
    '주소': addresses
})

In [11]:
seoul_starbucks_df['구'] = seoul_starbucks_df['주소'].str.extract(r'서울특별시\s(.*?구)')


In [12]:
seoul_starbucks_df

매장명                                              주소    구
0    역삼아레나빌딩                서울특별시 강남구 언주로 425 (역삼동)1522-3232  강남구
1     논현역사거리               서울특별시 강남구 강남대로 538 (논현동)1522-3232  강남구
2    신사역성일빌딩               서울특별시 강남구 강남대로 584 (논현동)1522-3232  강남구
3     국기원사거리               서울특별시 강남구 테헤란로 125 (역삼동)1522-3232  강남구
4     대치재경빌딩             서울특별시 강남구 남부순환로 2947 (대치동)1522-3232  강남구
..       ...                                             ...  ...
619     사가정역                      서울특별시 중랑구 면목로 3101522-3232  중랑구
620      상봉역                서울특별시 중랑구 망우로 307 (상봉동)1522-3232  중랑구
621       묵동  서울특별시 중랑구 동일로 952 (묵동, 로프트원 태릉입구역) 1층1522-3232  중랑구
622      양원역              서울특별시 중랑구 양원역로10길 3 (망우동)1522-3232  중랑구
623      중화역                   서울특별시 중랑구 봉화산로 35 1층1522-3232  중랑구

[624 rows x 3 columns]

In [120]:
seoul_starbucks_df.to_csv("starbucks_coffee_seoul.csv", index=False, encoding="utf-8-sig")

# 문제 2

In [13]:
# 웹 페이지 접속

url = 'https://www.ediya.com/contents/find_store.html'
driver = webdriver.Chrome()
driver.get(url)

In [14]:
# '주소' 버튼 클릭

find_address = driver.find_element(By.CSS_SELECTOR, '#contentWrap > div.contents > div > div.store_search_pop > ul > li:nth-child(2) > a')
find_address.click()

In [15]:
# 서울시 구 리스트 가져오기

gu_list = seoul_starbucks_df['구'].unique()
gu_list

array(['강남구', '강북구', '강서구', '관악구', '광진구', '금천구', '노원구', '도봉구', '동작구',
       '마포구', '서대문구', '서초구', '성북구', '송파구', '양천구', '영등포구', '은평구', '종로구',
       '중구', '강동구', '구', '동대문구', '성동구', '용산구', '중랑구'], dtype=object)

In [16]:
ediya_list = []

for gu in gu_list:
    # 검색창 초기화 및 구 이름 입력
    keyword = driver.find_element(By.ID, "keyword")
    keyword.clear()
    keyword.send_keys(f'서울 {gu}')
    
    # 검색 버튼 클릭
    search_button = driver.find_element(By.CSS_SELECTOR, '#keyword_div > form > button')
    search_button.click()
    time.sleep(2)  
    
    html = driver.page_source
    soup = BeautifulSoup(html, "html.parser")
    
    # 매장 리스트 추출
    ul_tag = soup.find("ul", id="placesList")
    if not ul_tag:  # 검색 결과가 없는 경우
        print(f"{gu}에 매장이 없습니다.")
        continue

    store_list = ul_tag.find_all("dl")
    
    # 각 매장 정보 추출
    for store in store_list:
        name = store.find("dt").get_text(strip=True)  # 매장명
        address = store.find("dd").get_text(strip=True)  # 주소
        
        # 구 정보 추출
        address_parts = address.split(" ")
        store_gu = address_parts[1] if len(address_parts) >= 2 else "구 정보 없음"
        
        # 데이터 저장
        ediya_list.append({
            "매장명": name,
            "주소": address,
            "구": store_gu
        })

# DataFrame 생성
seoul_ediya_df = pd.DataFrame(ediya_list)

# 데이터 확인
print(seoul_ediya_df)


             매장명                           주소    구
0        강남YMCA점                   서울 강남구 논현동  강남구
1    강남구청역아이티웨딩점  서울 강남구 학동로 338 (논현동, 강남파라곤)  강남구
2        강남논현학동점      서울 강남구 논현로131길 28 (논현동)  강남구
3          강남대치점   서울 강남구 역삼로 415 (대치동, 성진빌딩)  강남구
4          강남도산점      서울 강남구 도산대로37길 20 (신사동)  강남구
..           ...                          ...  ...
550       중랑교차로점         서울 중랑구 동일로 683 (면목동)  중랑구
551        중랑묵동점           서울 중랑구 공릉로 42 (묵동)  중랑구
552         중랑역점         서울 중랑구 망우로 198 (상봉동)  중랑구
553         중화동점       서울 중랑구 동일로129길 1 (중화동)  중랑구
554        화랑대역점  서울 중랑구 신내로25가길 2 (묵동, 현동학당)  중랑구

[555 rows x 3 columns]


In [32]:
seoul_ediya_df.to_csv("ediya_coffee_seoul.csv", index=False)

# 문제 3

## 위도 및 경도 변환
- Google Maps API를 사용하여 매장 주소를 위도와 경도로 변환.
- 이를 통해 매장 간의 물리적 거리를 계산 가능.

In [129]:
# pip install googlemaps

In [34]:
import pandas as pd
import googlemaps
from geopy.distance import geodesic
import folium

gmaps = googlemaps.Client(key="AIzaSyCIurBd9l7qbK3SwkuRxyt7RT8o55iR1Ew")

ediya_df = pd.read_csv("C:\my_ws\과제\스타벅스 이디야 매장 데이터 분석\ediya_coffee_seoul.csv")  # 이디야 매장 데이터
starbucks_df = pd.read_csv("C:\my_ws\과제\스타벅스 이디야 매장 데이터 분석\starbucks_coffee_seoul.csv")  # 스타벅스 매장 데이터

# Google Maps API를 사용해 위도/경도 변환
def get_lat_lon_google(address):
    try:
        geocode_result = gmaps.geocode(address)
        if geocode_result:
            location = geocode_result[0]['geometry']['location']
            return (location['lat'], location['lng'])
        else:
            return (None, None)
    except Exception as e:
        print(f"Error with address '{address}': {e}")
        return (None, None)

# 이디야 매장 위도/경도 추가
ediya_df[['위도', '경도']] = ediya_df['주소'].apply(lambda x: pd.Series(get_lat_lon_google(x)))

# 스타벅스 매장 위도/경도 추가
starbucks_df[['위도', '경도']] = starbucks_df['주소'].apply(lambda x: pd.Series(get_lat_lon_google(x)))

# 데이터 확인
print(ediya_df.head())
print(starbucks_df.head())

           매장명                           주소    구         위도          경도
0      강남YMCA점                   서울 강남구 논현동  강남구  37.513679  127.031712
1  강남구청역아이티웨딩점  서울 강남구 학동로 338 (논현동, 강남파라곤)  강남구  37.516551  127.040139
2      강남논현학동점      서울 강남구 논현로131길 28 (논현동)  강남구  37.515190  127.027554
3        강남대치점   서울 강남구 역삼로 415 (대치동, 성진빌딩)  강남구  37.501434  127.052328
4        강남도산점      서울 강남구 도산대로37길 20 (신사동)  강남구  37.522282  127.031480
       매장명                                   주소    구         위도          경도
0  역삼아레나빌딩     서울특별시 강남구 언주로 425 (역삼동)1522-3232  강남구  37.501109  127.042893
1   논현역사거리    서울특별시 강남구 강남대로 538 (논현동)1522-3232  강남구  37.510248  127.022377
2  신사역성일빌딩    서울특별시 강남구 강남대로 584 (논현동)1522-3232  강남구  37.513958  127.020663
3   국기원사거리    서울특별시 강남구 테헤란로 125 (역삼동)1522-3232  강남구  37.499615  127.031676
4   대치재경빌딩  서울특별시 강남구 남부순환로 2947 (대치동)1522-3232  강남구  37.494525  127.062584


## 거리 계산
- 이디야 매장에서 가장 가까운 스타벅스 매장까지의 거리를 계산.
- `geopy` 라이브러리의 `geodesic()` 함수를 사용하여 두 좌표 간의 거리(미터 단위)를 계산.

In [ ]:
from geopy.distance import geodesic

# 거리 계산 함수
def calculate_distance(lat1, lon1, lat2, lon2):
    if None in [lat1, lon1, lat2, lon2]:
        return None
    return geodesic((lat1, lon1), (lat2, lon2)).meters

# 각 이디야 매장별로 가장 가까운 스타벅스 거리 계산
closest_distances = []
for _, ediya_row in ediya_df.iterrows():
    ediya_location = (ediya_row['위도'], ediya_row['경도'])
    distances = []
    for _, starbucks_row in starbucks_df.iterrows():
        starbucks_location = (starbucks_row['위도'], starbucks_row['경도'])
        distance = calculate_distance(*ediya_location, *starbucks_location)
        distances.append(distance)
    closest_distances.append(min(distances) if distances else None)

# 결과 추가
ediya_df['가장 가까운 스타벅스 거리(m)'] = closest_distances
ediya_df['500m 이내 여부'] = ediya_df['가장 가까운 스타벅스 거리(m)'] <= 500

# 확인
print(ediya_df[['매장명', '주소', '가장 가까운 스타벅스 거리(m)', '500m 이내 여부']])

             매장명                           주소  가장 가까운 스타벅스 거리(m)  500m 이내 여부
0        강남YMCA점                   서울 강남구 논현동         139.043511        True
1    강남구청역아이티웨딩점  서울 강남구 학동로 338 (논현동, 강남파라곤)          97.664731        True
2        강남논현학동점      서울 강남구 논현로131길 28 (논현동)         286.536961        True
3          강남대치점   서울 강남구 역삼로 415 (대치동, 성진빌딩)         278.953618        True
4          강남도산점      서울 강남구 도산대로37길 20 (신사동)         263.783825        True
..           ...                          ...                ...         ...
550       중랑교차로점         서울 중랑구 동일로 683 (면목동)         500.187004       False
551        중랑묵동점           서울 중랑구 공릉로 42 (묵동)         179.013222        True
552         중랑역점         서울 중랑구 망우로 198 (상봉동)          19.216059        True
553         중화동점       서울 중랑구 동일로129길 1 (중화동)         551.709412       False
554        화랑대역점  서울 중랑구 신내로25가길 2 (묵동, 현동학당)         430.389925        True

[555 rows x 4 columns]


## 시각화
- 이디야 : 파란색, 스타벅스 : 초록색
- Folium을 사용해 이디야와 스타벅스 매장을 서울 지도에 표시.
- 시각적으로 매장 간의 위치 관계를 확인.

In [38]:
# 서울 중심 좌표
seoul_map = folium.Map(location=[37.5665, 126.9780], zoom_start=12)

# 이디야 매장 추가
for _, row in ediya_df.iterrows():
    folium.Marker(
        location=[row['위도'], row['경도']],
        popup=f"이디야: {row['매장명']} ({row['가장 가까운 스타벅스 거리(m)']}m)",
        icon=folium.Icon(color="blue", icon="coffee"),
    ).add_to(seoul_map)

# 스타벅스 매장 추가
for _, row in starbucks_df.iterrows():
    folium.Marker(
        location=[row['위도'], row['경도']],
        popup=f"스타벅스: {row['매장명']}",
        icon=folium.Icon(color="green", icon="coffee"),
    ).add_to(seoul_map)

# 지도 저장
seoul_map.save("seoul_coffee_map.html")
print("지도 저장 완료: seoul_coffee_map.html")


지도 저장 완료: seoul_coffee_map.html


In [39]:
seoul_map

## 분석 기준

- 고객은 일반적으로 도보로 약 **5~10분(약 500m)** 이내의 거리를 "가까움"으로 인식합니다.
- **국내 상권 분석**에서는 **500m 반경**을 주요 **1차 상권**으로 정의하며, 매장 간의 유동 인구를 분석하는 데 활용합니다.  
  - 출처: 소상공인시장진흥공단 상권정보시스템, 서울시 상권분석 서비스
- **유통 연구**에 따르면, 고객이 가장 선호하는 매장 간 거리는 **300~500m**입니다.  
  - 출처: Clarke (2000), *Retail Location and Consumer Behavior*

따라서, **500m 이내**는 매장 간 "의도적 가까움"을 판단하기에 적절한 기준입니다.

- 500m 이내의 근접성 : 이디야 매장과 가장 가까운 스타벅스 매장이 **500m 이내**에 있는지를 기준으로 근접성을 분석
- 평균 거리 : 매장 간 평균 거리를 계산하여 전반적인 위치 관계를 파악
- 500m 이상 떨어진 매장 : 독립적인 위치 전략을 가진 매장의 비율을 분석

In [40]:
# 500m 이내 비율
within_500m_ratio = ediya_df['500m 이내 여부'].mean() * 100
print(f"이디야 매장의 {within_500m_ratio:.2f}%가 스타벅스 매장 500m 이내에 위치합니다.")

# 거리 통계
print(ediya_df['가장 가까운 스타벅스 거리(m)'].describe())

# 500m 이상 떨어진 매장 수
far_from_starbucks = ediya_df[ediya_df['500m 이내 여부'] == False]
print(f"500m 이상 떨어진 이디야 매장 수: {len(far_from_starbucks)}")


이디야 매장의 68.65%가 스타벅스 매장 500m 이내에 위치합니다.
count     555.000000
mean      407.992801
std       354.755051
min         0.000000
25%       144.568543
50%       303.459255
75%       591.179449
max      2012.114713
Name: 가장 가까운 스타벅스 거리(m), dtype: float64
500m 이상 떨어진 이디야 매장 수: 174


# 분석 결과

- 500m 이내 비율 : 전체 이디야 매장 중 **68.65**%가 스타벅스 매장 **500m 이내**에 위치


- 평균 거리 : 이디야 매장과 가장 가까운 스타벅스 매장 간 **평균 거리는 약 408m**


- 500m 이상 거리 : 약 **31.35**%의 이디야 매장은 스타벅스 매장과 **500m 이상** 떨어져 있음

# **결론**

---

## **입지 전략**
- **이디야 커피 매장**은 대부분 **스타벅스 매장 근처**에 위치하여, **고객 트래픽을 공유하려는 전략**을 사용하는 것으로 보입니다.
- 이러한 전략은 **스타벅스의 강력한 브랜드 인지도**를 활용하여, 이디야가 **유사한 고객층을 유치**하려는 의도로 해석될 수 있습니다.

---

## **독립 전략**
- 약 **31.5**%의 이디야 매장은 **스타벅스와 거리가 멀리 떨어져** 있어, **독립적인 입지 전략**을 추구하는 것으로 보입니다.
- 이러한 매장은
  - 경쟁이 적은 지역에서 **로컬 고객층을 확보**하거나,
  - 특정 상권에서 **독립적으로 성공하려는 전략**으로 해석됩니다.

---

## **전략적 위치 선정**
- **이디야 커피**는 두 가지 전략을 혼합적으로 활용하고 있습니다
  1. **스타벅스와의 근접성을 활용한 전략**  
     - 트래픽이 많은 지역에서 스타벅스 매장과 가까운 위치를 선정하여, 고객 유입 효과를 극대화.
  2. **독립적 입지를 통한 상권 확보 전략**  
     - 경쟁이 적은 지역에서 고유한 입지 전략을 통해 독자적인 고객층 확보.

- 이러한 접근 방식은 **다양한 고객층**을 확보하려는 **혼합 전략**의 일환으로 보입니다.


# 추가 분석 가능성
- 거리 외에도 매장 매출, 방문자 수 등 데이터가 있다면, 입지 전략의 **경제적 효과**를 분석할 수 있습니다.
- 특정 지역(예: 강남, 홍대 등) 내 **경쟁 강도와 브랜드 인지도**를 비교 분석하면, 두 브랜드의 **세부적인 시장 전략**을 더 잘 이해할 수 있습니다.